File: GenericTesting/WPFControls/Charting/PlotPoint.vb
Public Class PlotPoint(Of T)
  Inherits PlotPoint

  Public Sub New(point As T)
    If Not (GetType(T) Is GetType(Decimal) Or GetType(T) Is GetType(Double) Or GetType(T) Is GetType(Integer) Or GetType(T) Is GetType(Date) Or GetType(T) Is GetType(DateTime)) Then Throw New ArgumentException("Cannot have a plot point that is not Decimal, Double, Integer, Date or DateTime")

    Me.Point = point
  End Sub
  Public Property Point As T

  Public Overrides ReadOnly Property PointType As Type
    Get
      Return GetType(T)
    End Get
  End Property
End Class

Public MustInherit Class PlotPoint
  MustOverride ReadOnly Property PointType As Type
End Class

File: GenericTesting/GenericTesting/Models/Charting/PlotPoint.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Controls.Charting
{
  public class PlotPoint<T> : PlotPoint
  {
    public PlotPoint(T point)
    {
      var typ = typeof(T);
                             
      if (!(typeof(DateTime) == typ || typeof(double) == typ || typeof(decimal) == typ || typeof(int) == typ))
        throw new ArgumentException("Cannot have a plot point that is not Decimal, Double, Integer, Date or DateTime");

      this.Point = point;
    }
    public T Point { get; set; }

    public override Type PointType
    {
      get { return typeof(T); }
    }
  }

  public abstract class PlotPoint
  {
    public abstract Type PointType { get; }
  }
}
